### ISSUE TITLE
borg mount: Should file content be cached on (local) disk?

### ISSUE THREAD

**Issue by textshell** [borgbackup/borg#2662]
#2572 removed caching of file content because in many cases the VFS should do a better job than borg caching to disk anway.
But there are use cases like mounting a filesystem image / block device backup via `borg mount`, a loopback device and various filesystem/block devices that can easily have a working set much bigger than what the VFS can cache.
So it's not really clear if caching should be done or not.

**Comment by ThomasWaldmann** [comment]
if it can't be decided, make it an option?

**Comment by textshell** [comment]
yes, making it an option would of course allow for both ways. Likely for this there is no one size fits all solution.

**Comment by enkore** [comment]
/tmp, which Borg will use by default, is often a tmpfs nowadays. If the machine has no swap, then anything written to it will use up pages from the page cache.

> But there are use cases like mounting a filesystem image / block device backup via borg mount, a loopback device and various filesystem/block devices that can easily have a working set much bigger than what the VFS can cache.

I don't think conflating issues will do any good here.
Mounting a file system through a loop device always has the drawback of double caching (seems to be fixed in some combinations since Linux 4.4), which is independent of whatever caching borg.fuse might do, apart from having another data cache would mean triple caching or double caching, respectively.
Since a file system uses the VFS, it always use the page cache for data and inode/directory caches and some others (depending on the file system).

Now, for the issue at hand.

> have a working set much bigger than what the VFS can cache.

Use bcache. I don't think borg.fuse will ever do any better than that, and as was already established, for these special cases special configuration is required anyway.

**Comment by enkore** [comment]
It's also notable to mention that borg.fuse already brings it's own inode cache.

**Comment by textshell** [comment]
I'm unconcerned with double or triple caching. That might be suboptimal, but i don't think it's too bad.
Also i don't count an on disk cache as duplicating an in memory cache, because memory is quite a bit smaller than disk in most cases.
The bcache suggestion might work, but i think it's impractical to expect our users to assemble complicated block device stacks on top of borg when borg already has a caching facility.
The problem here is that if you consider all aspects those stacks would grow quite a lot. If you need caching externally in addition to the loopback and possibly make the device 'writable' for journaling file systems to be happy (journal replay on start is quite common) users end up with stacks that nobody can remember the setup of:
```
borg mount -> losetup -> bcache -> dmsetup snapshot -> mount
```
Yes, removing bcache is not that much, but it's something that is easy to do for usability.
Also even if i don't have a good example i think there are also non disk-image use cases where some peace of userspace might repeatedly need to read the same data. Maybe while grepping around with multiple tries to find the file that needs to be restored.

**Comment by enkore** [comment]
I do not agree with your sentiment that clogging the primary caches is a ok thing to do (and then bolt on some other caching to work around it?).

You also seem to be missing the not-so-unimportant detail that Borg stopped caching the reads in your examples over one year ago.

borg.fuse performs poorly in some circumstances (btw. until I fixed it it performed very poorly in almost all circumstances, especially sequential reads). These are not solved by doing random changes and seeing whether they have an effect. Reproduce, analyse, devise fix, implement fix. Then that fix might be doing caching differently or adding another cache.

**Comment by textshell** [comment]
I tried both now. (tangent: The change triggering this ticket should not have made any difference as single gets are not cached anyway)
The system might be a more pessimistic test setup, but it's my main system so i think it is a realistic setup. My test image is a VM image with a mostly out of the box installed debian without much additional packages (mostly an empty VM)
(spinning rust, DSL with 10Mbit/s, RAM mostly over full with stuff)
```
Without caching:
# time borg mount -o allow_other borg@server:vm-test::start ~/mnt/borg/
real    0m2.131s
# time (losetup -fv ~mnt/borg/test.img; kpartx -a /dev/loop0; while [ ! -e /dev/mapper/loop0p1 ] ; do sleep .1; done; mount /dev/mapper/loop0p1 /mnt)
real    0m38.883s
# time find /mnt/ | wc -l
21418
real    4m8.877s
# time find /mnt/ | wc -l
21418
real    0m0.033s
With caching: data = self.key.decrypt(id, list(self.repository.get_many([id]))[0])

# time borg mount -o allow_other borg@server:vm-test::start ~/mnt/borg/
real    0m3.097s
# time (losetup -fv ~mnt/borg/test.img; kpartx -a /dev/loop0; while [ ! -e /dev/mapper/loop0p1 ] ; do sleep .1; done; mount /dev/mapper/loop0p1 /mnt)
real    0m35.522s
# time find /mnt/ | wc -l
21418
real    1m19.554s
# time find /mnt/ | wc -l
21418
real    0m0.033s

image stats: dev/mapper/loop0p1  9.3G  698M  8.1G   8% /mnt
```

btw, bcache needs special formating for the backing store. And dm-cache seems to be quite complicated for users to "just remember how to use"

**Comment by enkore** [comment]
I assume you tested this with my new cache rather than the previous implementation?

**Comment by textshell** [comment]
I used git master (b51198ebe7) so yes, with the new cache and enough space (64G free on that filesystem) for the cache.

**Comment by enkore** [comment]
I'd suggest to check how the previous implementation behaved.

**Comment by ThomasWaldmann** [comment]
1.4-maint: no change planned

master: borg2 will do some caching in borgstore, we'll have to evaluate this again with some of the next betas.
